### PR TITLE
feat: Add trigger to combobox #1136

### LINKED ISSUE #1136: Add trigger to combobox
via @azim-b

### PR DESCRIPTION
**Pull request by aalencar**
* Add prop trigger to combobox allowing the choice to be sent to the server right away when it is selected.

resolves #1136

### CODE REVIEW

File: py/examples/combobox.py [hunk 1]
```diff
@@ -24,6 +25,8 @@ async def serve(q: Q):
                         choices=combobox_choices, disabled=True),
             ui.combobox(name='combobox_error', label='Enter or choose a color', placeholder='Color...', value='Blue',
                         choices=combobox_choices, error='This combobox has an error!'),
+            ui.combobox(name='combobox_trigger', trigger=True, label='Enter or choose a color (trigger on selection)', placeholder='Color...', value=q.args.combobox_trigger,
```
Comment (aalencar) [inline py/examples/combobox.py:28]: @mturoci is there something wrong with using the value that was recently send to the server? And how do I conditionally pass arguments in Python like in JS `value || fallback`?

File: py/examples/picker.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ async def serve(q: Q):
         ]
     else:
         q.page['example'] = ui.form_card(box='1 1 4 10', items=[
-            ui.picker(name='picker', label='Place an order (try Spam, Eggs or Ham):', choices=[
+            ui.picker(name='picker', trigger=True, label='Place an order (try Spam, Eggs or Ham):', choices=[
```
Comment (lo5) [inline py/examples/picker.py:17]: This example is unrelated to the combobox. Is this an intentional change?

File: py/examples/combobox.py (hunk 1 above)
Comment (lo5) [reply py/examples/combobox.py:28]: It's confusing to read. It's enough if the example line demonstrates how to make a combobox submit its value when changed. The `value=q.args.combobox_trigger` is unnecessary.

Comment (lo5) [inline py/examples/combobox.py:28]:
```suggestion
            ui.combobox(name='combobox_trigger', trigger=True, label='Enter or choose a color (submit color when changed)', placeholder='Color...', value='Blue',
```

**Review by lo5 — CHANGES_REQUESTED** [review]

File: py/examples/combobox.py (hunk 1 above)
Comment (mturoci) [reply py/examples/combobox.py:28]: >is there something wrong with using the value that was recently send to the server

`value` is meant to be "initial" value - all wave components should be uncontrolled. That means the `value` would work only if you recreate the card.

>how do I conditionally pass arguments in Python 

`value or fallback`

Comment (mturoci) [reply py/examples/combobox.py:28]: Anyway it's better to create a separate example for `trigger`, like we have for `textbox` for example. The reason is that triggering will submit also filled values from other comboboxes.

File: py/examples/picker.py (hunk 2 above)
Comment (aalencar) [reply py/examples/picker.py]: I was checking the picker behavior on trigger and commited it by mistake. Reverted, thanks.

File: py/examples/combobox.py (hunk 1 above)
Comment (aalencar) [reply py/examples/combobox.py]: I added a new example like we have for textbox

File: py/examples/combobox.py [hunk 3]
```diff
@@ -14,6 +14,7 @@ async def serve(q: Q):
             ui.text(f'combobox={q.args.combobox}'),
             ui.text(f'combobox_disabled={q.args.combobox_disabled}'),
             ui.text(f'combobox_error={q.args.combobox_error}'),
+            ui.text(f'combobox_trigger={q.args.combobox_trigger}'),
```
Comment (mturoci) [inline py/examples/combobox.py:17]: out

File: ui/src/combobox.tsx [hunk 4]
```diff
@@ -51,6 +51,8 @@ export interface Combobox {
   visible?: B
   /** An optional tooltip message displayed when a user clicks the help icon to the right of the component. */
   tooltip?: S
+  /** True if the choice should be submitted when the dropdown value changes. */
```
Comment (mturoci) [inline ui/src/combobox.tsx:54]: Not just dropdown.

**Review by mturoci — CHANGES_REQUESTED** [review]
Thanks @aalencar. Don't forget that combobox also allows free text (not just dropdown) which needs to be triggered as well. The free text needs to be debounced so that we do not submit every keystroke. See `textbox` for an implementation reference.

File: py/examples/combobox.py (hunk 3 above)
Comment (aalencar) [reply py/examples/combobox.py]: Removed.

File: ui/src/combobox.tsx (hunk 4 above)
Comment (aalencar) [reply ui/src/combobox.tsx]: Improved.

**Comment by aalencar** [comment]
> The free text needs to be debounced so that we do not submit every keystroke

I implemented the debounce but I noticed that `onChange` is only triggered on blur, pressing enter,  or selecting a choice.
Based on that, there is no need to debounce if we prevent the user from sending undefined choices. I added a guard to prevent that.

**Comment by mturoci** [comment]
>I implemented the debounce but I noticed that onChange is only triggered on blur, pressing enter, or selecting a choice.

Good to know.

>Based on that, there is no need to debounce if we prevent the user from sending undefined choices. I added a guard to prevent that.

I don't understand this one. The value will never be undefined thanks to `const v = option?.text || value || ''`, right?

**Comment by aalencar** [comment]
> > I implemented the debounce but I noticed that onChange is only triggered on blur, pressing enter, or selecting a choice.
> 
> Good to know.
> 
> > Based on that, there is no need to debounce if we prevent the user from sending undefined choices. I added a guard to prevent that.
> 
> I don't understand this one. The value will never be undefined thanks to `const v = option?.text || value || ''`, right?

By undefined I mean some choice that is not present in choices / dropdown (i.e. not defined by the user).

Example: `choices = ['red', 'yellow', 'green']`, if user types `blue` or `yello` it won't send it to the server

 I made it more confusing by using the word 'undefined' in a javascript context. I hope it's more clear now.

**Comment by mturoci** [comment]
But this is not the correct logic, combobox allows free text as well which should be sent to server also. Otherwise, there would be no difference between combobox and dropdown basically. Please revert the condition.

**Comment by aalencar** [comment]
> But this is not the correct logic, combobox allows free text as well which should be sent to server also. Otherwise, there would be no difference between combobox and dropdown basically. Please revert the condition.

Reverted.

**Review by mturoci — APPROVED** [review]
Thanks @aalencar!
